Add container selection dropdown(#2551)

chentong7 opened the pull request:
## Description

One design requirement we have is to support multiple Fluid Containers. From the perspective of the debugger object tooling, consumers may initialize a debugger for as many containers as their application has. These get stored off in a "registry", which is implemented as a map from container ID strings to any associated debugger object.

See the code here for more details: [FluidFramework/Registry.ts at dev/client-debugger-prototype · microsoft/FluidFramework (github.com)](https://github.com/microsoft/FluidFramework/blob/dev/client-debugger-prototype/packages/tools/client-debugger/src/Registry.ts)

## Example
![0](https://user-images.githubusercontent.com/114451900/207487203-79e43722-e00d-423f-9378-51d109592b9a.png)
![1](https://user-images.githubusercontent.com/114451900/207487214-fdf79681-57a0-4d02-a575-27e48b15d3ea.png)

Josmithr committed:
fix: Type tests

Josmithr committed:
fix: Type tests

Josmithr committed:
Merge branch 'main' into dev/client-debugger-prototype

Josmithr committed:
build: Update dependencies

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' into debugger-object-prototype

Josmithr committed:
test: Make test scripts fluid-build compliant

Josmithr committed:
test: Fix test config

Josmithr committed:
test: Fix mocha config

Josmithr committed:
tools: Make build scripts fluid-build compliant

Josmithr committed:
style: Prettier

Josmithr committed:
fix: Event name

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' into debugger-object-prototype

Josmithr committed:
[WIP] Start using debug object in view

Josmithr committed:
style: Prettier

Josmithr committed:
fix: Add missing export

Josmithr committed:
test: Add test infra for accessing debug object

Josmithr committed:
refactor: Disable export pattern linter rules while prototyping

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' into debugger-object-prototype

Josmithr committed:
build: Make fluid-static a dev dep so we don't actually use its types

Josmithr committed:
improvement: Add more data and events to debugger

Josmithr committed:
refactor: Add some common props interfaces

Josmithr committed:
refactor: Add some helper React hooks for using data derived from the debugger

Josmithr committed:
remove: Obsolete helper function

Josmithr committed:
refactor" [WIP] Use debugger

Josmithr committed:
refactor: Rename method and update docs

Josmithr committed:
refactor: Rename methods and update docs

Josmithr committed:
feat: Add more data and events to debugger

Josmithr committed:
docs: Add TODO

Josmithr committed:
refactor: Rename method and re-order methods

Josmithr committed:
feat: Add manual user action methods to debugger

Josmithr committed:
refactor: [WIP]: Use debug object

chentong7 committed:
Track audience change history and show it (#1956) (#12859)

## Description

Currently, the AudienceView component simply displays the current
audience members. It would be useful to also track changes over time.

By default, we probably want to keep this view turned off, so maybe it
belongs under a drop-down (accordion). It could also be a modal dialogue
if that would allow the contents to fit better.

## Breaking Changes

> Listen on "memberAdded" and "memberRemoved" events and record this
information on history.

## Sample
![Audience History
View](https://user-images.githubusercontent.com/114451900/200736149-c21d01b1-921b-4850-998f-f28c79deaea9.png)

Josmithr committed:
docs: Expand comment

Josmithr committed:
feat: Add event notifying of debugger disposal, and add container data

Josmithr committed:
refactor: Split modules

Josmithr committed:
refactor: Use debugger

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' into debugger-object-prototype

Josmithr committed:
fix: Event emitter leak

Josmithr committed:
fix: Element key

Josmithr committed:
refactor: Only expose a single event hook for audience member changes

Josmithr committed:
docs: Expand documentation

Josmithr committed:
fix: Correctly signal and handle attach state

Josmithr committed:
refactor: Simplify attach and connection state exposure / events and update docs

Josmithr committed:
build: Add package scaffolding for chrome extension library

Josmithr committed:
docs: Fill in comment

Josmithr committed:
docs: Add missing copyright notices

Josmithr committed:
Merge branch 'debugger-object-prototype' into debugger-browser-extension-prototype

Josmithr committed:
style: Prettier

Josmithr committed:
build: Remove unneeded step in build script

Josmithr committed:
Merge branch 'debugger-object-prototype' into debugger-browser-extension-prototype

Josmithr committed:
refactor: Rename internal accessor methods

Josmithr committed:
Add prototype debugger object library (#12872)

NOTE: this is a PR into our development branch and not into main.

Adds a debugger library, and updates the viewer prototype to use it.

Requires adding a couple of internal-only methods to extract lower-level
data types: 1 on `IFluidContainer` and one on `IServiceAudience` to
allow entrypoint into the debugger library from the context of a
consumer using our 3rd party API surface. Later, we will expose an
adapter library that does this transparently, rather than requiring
calls to these methods.

Josmithr committed:
improvement: Clean up build and file naming

Josmithr committed:
refactor: Remove unneeded audience internal accessor

Josmithr committed:
fix: Event handler bindings leak

Josmithr committed:
docs: Remove TODO

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' into debugger-browser-extension-prototype

Josmithr committed:
improvement: Update icon

Josmithr committed:
fix: Get hello world working and fix icon format

Josmithr committed:
feat: Toggle view via extension button

Josmithr committed:
fix: Error statements

Josmithr committed:
Merge branch 'main' into dev/client-debugger-prototype

Josmithr committed:
remove: Ops data from debugger

Josmithr committed:
style: Run updated prettier config

Josmithr committed:
[WIP] Listener pattern

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' into debugger-browser-extension-prototype

Josmithr committed:
refactor: Move event handlers

Josmithr committed:
style: Prettier

Josmithr committed:
tools: Add "endregion" to spelling dictionary

Josmithr committed:
Revert "refactor: Move event handlers"

This reverts commit ff6275c1f27f0e49ca04a6862f6f7436d5f2c644.

Josmithr committed:
refactor: Move props interface

Josmithr committed:
feat: Add debugger state summarization

Josmithr committed:
build: Add needed deps

Josmithr committed:
feat: Debugger extension prototype

Josmithr committed:
docs: Add comments

Josmithr committed:
refactor: Better type-safety

Josmithr committed:
fix: Type handling in FluidObjectView

Josmithr committed:
remove: Unneeded prototype code

Josmithr committed:
build: Update deps and linter configs

Josmithr committed:
test: Add test infra

Josmithr committed:
test: Restructure test infra

Josmithr committed:
test: Update test infra

Josmithr committed:
tools: Add code coverage scripts and cleanup related scripts and ignore files

Josmithr committed:
remove: Test webpack config

Josmithr committed:
remove: package-lock

Josmithr committed:
build: Add dev dep

Josmithr committed:
refactor: Move script un-binding to upper layer of script hierarchy

Josmithr committed:
test: Add tests

Josmithr committed:
refactor: Split script code into its own module to make testing easier

Josmithr committed:
fix: Test

Josmithr committed:
fix: Webpack config

Josmithr committed:
tools: Fix casing in webpack config

Josmithr committed:
refactor: Misc. cleanup

Josmithr committed:
feat: Add className constant for debug view

Josmithr committed:
test: Simplify tests

Josmithr committed:
test: Add scenario test

Josmithr committed:
improvement: Clean up appended scripts immediately

Josmithr committed:
fix: Coverage script

Josmithr committed:
tools: Omit test infra code from code coverage reporting

Josmithr committed:
tools: Add "Tinylicious" to spelling dictionary

Josmithr committed:
refactor: Reorganize test init

Josmithr committed:
docs: Add notes to the README

Josmithr committed:
refactor: Rename scripts and update docs

Josmithr committed:
style: Prettier

Josmithr committed:
revert: Test change

Josmithr committed:
tools: Update spelling dictionary

Josmithr committed:
tools: Update jest related infra

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' into debugger-browser-extension-prototype

Josmithr committed:
chore: Remove unneeded gitignore entry

Josmithr committed:
tools: Fix prettier ignore

Josmithr committed:
chore: Remove unneeded entry from gitignore

Josmithr committed:
tools: Add chai eslint plugin

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' into debugger-browser-extension-prototype

Josmithr committed:
fix: Debugger disposal notification

Josmithr committed:
fix: Coverage scripts

Josmithr committed:
refactor: Move FluidClientDebuggerProps to Registry.ts

Josmithr committed:
tools: Update prettier ignore

Josmithr committed:
refactor: Don't expose debugger types publicly

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' into debugger-browser-extension-prototype

Josmithr committed:
tools: Update .npmignore files

chentong7 committed:
Expose entrypoint of container

chentong7 committed:
Merge branch 'dev/client-debugger-prototype' of https://github.com/microsoft/FluidFramework into dev/client-debugger-prototype

Josmithr committed:
fix: Test helper

chentong7 committed:
Merge branch 'dev/client-debugger-prototype' of https://github.com/microsoft/FluidFramework into dev/client-debugger-prototype

Josmithr committed:
style: Prettier

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' into debugger-browser-extension-prototype

Josmithr committed:
tools: Re-add package-lock to ignore files (just in case they get generated locally)

Josmithr committed:
docs: Update remarks and add TODO

Josmithr committed:
feat: Add needed (internal) exports

Josmithr committed:
fix: Text syntax

Josmithr committed:
fix: More-correct type handling for data views

Josmithr committed:
style: Sort exports

Josmithr committed:
Merge remote-tracking branch 'upstream/dev/client-debugger-prototype' into debugger-browser-extension-prototype

Josmithr committed:
Merge remote-tracking branch 'upstream/dev/client-debugger-prototype' into debugger-browser-extension-prototype

Josmithr committed:
docs: Update TODO

Josmithr committed:
Debugger browser extension prototype (#13028)

NOTE: This is a PR into our dev branch, not into main.

Adds a Chrome extension package. The extension itself is pretty simple -
it just toggles the debugger view on/off.


![image](https://user-images.githubusercontent.com/54606601/202822209-99261d47-ca9f-4ad2-ad9f-9c134d39eebf.png)

Josmithr committed:
build: Update lock file

Josmithr committed:
style: Prettier

Josmithr committed:
fix: Import

Josmithr committed:
tools: Add package-lock to ignore files

Josmithr committed:
tools: Update package details to remove npm libarary package aspects

Josmithr committed:
style: Sort package.json contents

Josmithr committed:
refactor: Make modules non-tsx

Josmithr committed:
docs: Remove obsolete TODO

Josmithr committed:
build: Lock updates

chentong7 committed:
Add client debugger render

chentong7 committed:
Merge branch 'dev/client-debugger-prototype' of https://github.com/microsoft/FluidFramework into dev/client-debugger-prototype

chentong7 committed:
Revert "Add client debugger render"

This reverts commit c7aca315a3f81f459915b06e8607ddbe9e024e73.

Josmithr committed:
refactor: Update visualier to use new pattern

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' of https://github.com/microsoft/FluidFramework into dev/client-debugger-prototype

Josmithr committed:
refactor: More-correct typing

Josmithr committed:
remove: html file and handling, since it is ignored by fluid webpack infra anyways

chentong7 committed:
Add React-agnostic component entrypoint (#13132)

## Description

> Currently, the library exposes an entrypoint that can only be used in
a React context. We should expose a React-agnostic entrypoint that takes
in the ID of a DOM element, and does the work of mounting the rendered
component into that DOM element.

## Example
Before app container connected
![Screenshot
1](https://user-images.githubusercontent.com/114451900/204886678-29bd01c9-759b-46d3-a7e0-311a57c6997c.png)
After
![Screenshot
2](https://user-images.githubusercontent.com/114451900/204886729-2881d7fe-f268-45af-a2b1-8da310c17959.png)

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' of https://github.com/microsoft/FluidFramework into dev/client-debugger-prototype

Josmithr committed:
Merge branch 'main' into dev/client-debugger-prototype

Josmithr committed:
Merge branch 'main' into dev/client-debugger-prototype

Josmithr committed:
docs: Fill in packageDocumentation comment

Josmithr committed:
tools: Require explicit release tags

Josmithr committed:
fix: Correctly mark some exports as internal

Josmithr committed:
tools: Require explicit release tag

Josmithr committed:
revert: Config changes

Josmithr committed:
docs: Update comment

Josmithr committed:
fix: Call `resolve` in error case

Josmithr committed:
refactor: Reduce API surface a bit

Josmithr committed:
refactor: Re-add elements to API surface as needed for Chrome extension (internal only)

Josmithr committed:
docs: Gen installation section of README

Josmithr committed:
docs: Update packageDocumentation comment

Josmithr committed:
docs: Expand README

Josmithr committed:
docs: More README updates

Josmithr committed:
fix: API report copy scripts

Josmithr committed:
style: Prettier

Josmithr committed:
docs: Update README and packageDocumentation comment

Josmithr committed:
docs: Update build instructions

Josmithr committed:
docs: Add build and test instructions to README

Josmithr committed:
feat: Add containerNickname prop to debugger

Josmithr committed:
refactor: Update constructor to take props type

chentong7 committed:
Make the debugger panel resizable (#13247)

## Description
Including npm re-resizable package to make debugger panel resizable by
dragging.

## Example
<img width="841" alt="1"
src="https://user-images.githubusercontent.com/114451900/205764316-8b6aa421-c0b6-48d7-9bc2-f2d65164eb3e.png">
<img width="852" alt="2"
src="https://user-images.githubusercontent.com/114451900/205764327-4fc688c2-fce0-4f02-a980-a36ecf582013.png">

Josmithr committed:
Merge branch 'main' into dev/client-debugger-prototype

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' of https://github.com/microsoft/FluidFramework into dev/client-debugger-prototype

chentong7 committed:
Update Chrome extension prototype to use React-agnostic entry-point(#2670) (#13250)

## Description

> Once we've added the React-agnostic entry-point in the visualizer
package, we should update the Chrome extension to take advantage of it.
<img width="798" alt="1"
src="https://user-images.githubusercontent.com/114451900/206047737-2fe3a512-926b-4947-81a0-30ad97e2760b.png">

<img width="797" alt="2"
src="https://user-images.githubusercontent.com/114451900/206047746-c5f680bb-d400-4161-a4fc-4fba161c418f.png">

chentong7 committed:
Add container selection dropdown

chentong7 committed:
Add container selection dropdown

jikim-msft committed:
Add visualization to client-debugger-view (#13313)

#### Description 

https://dev.azure.com/fluidframework/internal/_workitems/edit/2668/


#### Objective 

This PR adds visualization effect to `client-debugger-view` by adding
icons such as `+` and `-` to intuitively display connected and
disconnected members of the container.

Colors represent information: 
    - `lightgreen`: `added`
    - `orange`: `removed`

<img width="900" alt="image"
src="https://user-images.githubusercontent.com/111468570/206539556-94da068b-0063-4414-8c42-c55ff692b39e.png">

Josmithr committed:
refactor: Move Counter widget into tests

Josmithr committed:
fix: Data view

Josmithr committed:
build: Lock file update

Josmithr committed:
fix: Remove bad typetest comments

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' of https://github.com/microsoft/FluidFramework into dev/client-debugger-prototype

Josmithr committed:
Merge branch 'main' into dev/client-debugger-prototype

Josmithr committed:
merge: branch 'main' into dev/client-debugger-prototype

Josmithr committed:
Merge branch 'main' into dev/client-debugger-prototype

jikim-msft committed:
Add SharedCounter editor control to client-debug-visualizer package (#13350)

#### Description 

https://dev.azure.com/fluidframework/internal/_workitems/edit/2656/

#### Objective 

<img width="911" alt="image"
src="https://user-images.githubusercontent.com/111468570/207467001-72a3fbcb-4138-4f4c-9209-86ebdd5fdd0e.png">

This PR aims to add functionality so that the users can freely
manipulate the `sharedCounter` values by incrementing, decrementing or
by passing an arbitrary integer value in the input field that would
function as a delta value.


#### Notes 
- There should be some CSS works to make the visualization neater after
the logic for adding `sharedCounter` editor is approved.
- The current method of adjusting `sharedCounter.value` is by repeating
`sharedCounter.increment` for `num - value` number of times. We might
have to think about coming up with a more efficient solution or changing
the design of `sharedCounter` to allow applying an arbitrary integer
value in the future.

Josmithr committed:
Merge branch 'main' into dev/client-debugger-prototype

Josmithr committed:
build: Update eslint config deps

Josmithr committed:
Merge branch 'main' into dev/client-debugger-prototype

Josmithr committed:
Merge branch 'dev/client-debugger-prototype' of https://github.com/microsoft/FluidFramework into dev/client-debugger-prototype

chentong7 committed:
Add listen to client debugger list change

chentong7 committed:
Add listen to client debugger list change

Josmithr commented on packages/tools/client-debug-view/src/components/ContainerSelectionDropdown.tsx:
```diff
@@ -0,0 +1,96 @@
+			const selectedDebugger = clientDebuggers.find((c) => {
+				return c.containerId === (option.key as string);
+			}) as IFluidClientDebugger;
+			props.clientDebugger = selectedDebugger;
```
This won't update the container selection from the perspective of `ClientDebugView`. In order to get the parent container to change its selection based on the drop-down, we probably need to have `ClientDebugView` manage the list of debuggers and the current selection and update this component's props to take in the list of debuggers, the current selection, and an `onSelectionChange` (or something like that) callback to inform `ClientDebugView` when the selection changes.

chentong7 replied:
Adding "currentDisplayDebugger" in register and make it updating in debugger.tsx

Josmithr commented on packages/tools/client-debugger/package.json:
```diff
@@ -1,102 +1,102 @@
-    "tinylicious": "0.4.89251",
-    "typescript": "~4.5.5"
-  }
+	"name": "@fluid-tools/client-debugger",
```
Looks like these are just formatting changes. We should probably revert them. Running `npm run format` in this directory should fix things up.

chentong7 replied:
Updated after `npm run format` Thx!

Josmithr commented on packages/tools/client-debugger/src/Registry.ts:
```diff
@@ -45,6 +46,66 @@ export interface FluidClientDebuggerProps {
 	containerNickname?: string;
 }
 
+export interface DebuggerRegistryEvents extends IEvent {
```
We should document both of these interfaces. We should also add the `@internal` tag to both of these, as I don't expect external (outside of our repo) consumers to use these types.

Josmithr commented on packages/tools/client-debugger/src/Registry.ts:
```diff
@@ -45,6 +46,66 @@ export interface FluidClientDebuggerProps {
+export class DebuggerRegistry extends IEventProvider<DebuggerRegistryEvents> {
+	private readonly registeredDebuggers: Map<string, FluidClientDebugger> = new Map();
+
+	constructor() {
```
Nit: 
```suggestion
	public constructor() {
```

I would also recommend running `npm run lint` from inside the package directory to make sure we're satisfying our linter requirements.

Josmithr commented on packages/tools/client-debugger/src/Registry.ts:
```diff
@@ -45,6 +46,66 @@ export interface FluidClientDebuggerProps {
+		}
+	}
+
+	public getRegisteredDebuggers(): Map<string, IFluidClientDebugger> {
```
Rather than returning the underlying map, it might be nice to just return a readonly array of the current list. We can add methods to this type for getting a debugger by ID so that the external functions can still access them.

chentong7 replied:
Make sense! updated.

Josmithr commented on packages/tools/client-debug-view/src/components/ContainerSelectionDropdown.tsx:
```diff
@@ -0,0 +1,96 @@
+		getFluidClientDebuggers(),
+	);
+
+	const [clientDebuggerOptions, setClientDebuggerOptions] = React.useState<IDropdownOption[]>(
```
We shouldn't need to store these options as component state. Since they're strictly derived from the value of `clientDebuggers` above, we can just recompute this list each time `clientDebuggers` changes.

chentong7 replied:
Modified clientDebuggerOptions to be updated alone with clientDebuggers.

Josmithr reviewed (changes requested):
Looking good! There are a couple of issues that we will need to fix, but the direction is looking great!

One thing I would recommend is writing some unit/component tests for the multi-container/debugger case to make sure our component is working as expected.

It also might be worth updating the sandbox app to use multiple containers / debuggers so we can continually test this functionality with the app. One idea for how we could test this would be to have the app always create a fresh second container (one for which we don't update the page url) with another widget of some sort, and create a debugger for that. Let me know if you would like to chat about this 🙂

chentong7 committed:
update pr

chentong7 committed:
update pr

chentong7 commented on packages/tools/client-debugger/src/Registry.ts:
```diff
@@ -45,6 +46,66 @@ export interface FluidClientDebuggerProps {
+
+		const clientDebugger = new FluidClientDebugger(props);
+		this.registeredDebuggers.set(containerId, clientDebugger);
+		this.emit("debuggerRegistered", containerId, clientDebugger);
```
Updated.

Josmithr committed:
Merge branch 'main' into dev/client-debugger-prototype

Josmithr commented on packages/tools/client-debug-view/src/Debugger.tsx:
```diff
@@ -88,12 +88,13 @@ export function FluidClientDebugger(props: FluidClientDebuggerProps): React.Reac
 		<Resizable
 			style={{
-				position: "fixed",
+				position: "absolute",
```
Just curious: what were these changes needed for?

chentong7 replied:
Set position as `absolute` will allow view, like "container dropdown list", to place in front of debugger background view. The render order will base on `z-index`.

Josmithr replied:
Ooh, gotcha. Sounds good!

Josmithr committed:
refactor: Parent client debugger packages under common parent directory (#13421)

This change will hopefully help isolate these packages from the others,
and make future package-structure changes easier / better scoped.

This change also renames the "client-debug-view" package to
"client-debugger-view", to better align with the other packages.

Josmithr committed:
Merge branch 'main' into dev/client-debugger-prototype

Josmithr committed:
style: Prettier

Josmithr committed:
docs: Fix report name

Josmithr committed:
docs: Revert obsolete report changes

chentong7 commented:
> Looking good! There are a couple of issues that we will need to fix, but the direction is looking great!
> 
> One thing I would recommend is writing some unit/component tests for the multi-container/debugger case to make sure our component is working as expected.
> 
> It also might be worth updating the sandbox app to use multiple containers / debuggers so we can continually test this functionality with the app. One idea for how we could test this would be to have the app always create a fresh second container (one for which we don't update the page url) with another widget of some sort, and create a debugger for that. Let me know if you would like to chat about this 🙂

I'm planning to include the UTs in this [task](https://dev.azure.com/fluidframework/internal/_workitems/edit/2390/)'s code change. I'd add multiple containers for sandbox app.

chentong7 committed:
update PR

chentong7 committed:
update PR

chentong7 committed:
ran prettier

chentong7 committed:
ran prettier

chentong7 committed:
update PR

chentong7 committed:
update PR

Josmithr commented on packages/tools/client-debugger/src/Registry.ts:
```diff
@@ -45,6 +46,122 @@ export interface FluidClientDebuggerProps {
+}
+
+/**
+ * Contract for maintaining a global client debugger registry to store all registered client debugger and current
```
Nit: This layer in the system doesn't know anything about our visualizer, so I would recommend documenting this strictly in terms of what it does, and not how it is consumed by the visualizer.

Perhaps something like:

> Manages initialization and teardown of {@link IFluidClientDebugger}s, and notifies subscribers to changes via its {@link | events}.

chentong7 replied:
updated.

Josmithr commented on packages/tools/client-debugger/src/Registry.ts:
```diff
@@ -45,6 +46,122 @@ export interface FluidClientDebuggerProps {
+	/**
+	 * The debugger is currently displaying.
+	 */
+	private currentDisplayDebugger: IFluidClientDebugger;
```
This library really shouldn't know anything about our visualizer library. The visualizer code should be what manages the "current" debugger. This class should just fire off its 2 change events when appropriate, and our visualizer library should respond accordingly.

chentong7 replied:
Removed it

chentong7 committed:
Merge branch 'tong_pr' of https://github.com/chentong7/FluidFramework into tong_pr

chentong7 committed:
Merge branch 'main' of https://github.com/chentong7/FluidFramework into tong_pr

chentong7 committed:
merge main

Josmithr commented on packages/tools/client-debugger/src/Registry.ts:
```diff
@@ -0,0 +1,252 @@
+/*!
```
This file is showing as a new addition in the diff, but it should already exist. Have you merged main / rebased onto it?

chentong7 replied:
Yea. It sync-ed from latest main branch.

Josmithr replied:
The file still appears as being added here. You might need to manually delete it.

Josmithr commented on packages/tools/client-debugger/client-debugger-chrome-extension/src/test/ClientUtilities.ts:
```diff
@@ -0,0 +1,127 @@
+/*!
```
I would recommend waiting for #13604 to merge and pull those changes in. We will probably not need to add this file, and can instead use the Container mocks that PR adds.

chentong7 replied:
I see! Revert this change

Josmithr commented on packages/tools/client-debug-view/src/components/ContainerSelectionDropdown.tsx:
```diff
@@ -0,0 +1,84 @@
+
+/**
+ * Small header that displays core container data.
+ *
```
Nit
```suggestion
```

Josmithr commented on packages/tools/client-debug-view/src/components/ContainerSelectionDropdown.tsx:
```diff
@@ -0,0 +1,84 @@
+} from "@fluid-tools/client-debugger";
+
+/**
+ * Small header that displays core container data.
```
I don't think this is accurate. This component is a drop-down menu for selecting containers from a list.

chentong7 replied:
Fixed

Josmithr reviewed (changes requested):
I left a couple comments, mostly related to the way the `client-debugger` and `client-debugger-view` libraries are intended to interact.

chentong7 committed:
updates(2023)

chentong7 committed:
change to pull registry in Debugger.tsx level

chentong7 committed:
Merge branch 'main' of https://github.com/chentong7/FluidFramework into tong_pr

chentong7 committed:
update

chentong7 committed:
testing

chentong7 committed:
test in local

Josmithr commented on packages/tools/client-debugger/client-debugger/src/Registry.ts:
```diff
@@ -14,35 +16,103 @@ import { IFluidClientDebugger } from "./IFluidClientDebugger";
-     * debugger instances.
-     */
-    containerNickname?: string;
+	/**
```
Nit: Looks like there are some whitespace changes here. Be sure to run `npm run format` inside the package - should clean up whitespace diffs.

chentong7 replied:
run the format. Thanks for catching this up.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx:
```diff
@@ -32,48 +39,68 @@ export interface FluidClientDebuggerProps {
 
+    const [clientDebuggers, setClientDebuggers] = React.useState<IFluidClientDebugger[]>(getFluidClientDebuggers());
+
+    const [selectedContainerId, setSelectedContainerId] = React.useState<string>(
```
Instead of using an empty string as the default value here, I would recommend allowing this value to be `undefined` and having its default be `undefined`.
```suggestion
    const [selectedContainerId, setSelectedContainerId] = React.useState<string | undefined>(
```

chentong7 replied:
Fixed

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx:
```diff
@@ -32,48 +39,68 @@ export interface FluidClientDebuggerProps {
         clientDebugger?.disposed ?? false,
     );
 
+    const debuggerRegistry: DebuggerRegistry = getDebuggerRegistry();
```
We're already getting the full list of debuggers above. We probably want to move this higher up, and get the list of debuggers from this directly.

chentong7 replied:
I think monitor client debugger's change in here is reasonable. The higher up is `RenderClientDebugger`. Or should I have a debugger controller tsx to be the higher up call?

Josmithr replied:
I'm sorry, I completely missed your response here earlier.

I'm suggesting that since we're accessing the full list of debuggers on line 47 via `getFluidClientDebuggers` (which internally uses the registry), it might be nice to move this line above that, and replace the call to `getFluidClientDebuggers` with a call to `debuggerRegistry.getRegisteredDebuggers`.

This isn't a huge deal, but it would remove a lookup call on the window object, and I think it makes the logical flow of the function a little nicer.

chentong7 replied:
Ah, I see! Fixed it!

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx:
```diff
@@ -32,48 +39,68 @@ export interface FluidClientDebuggerProps {
+        getFirstDebugger()?.containerId ?? "",
+    );
+
     const [clientDebugger, setClientDebugger] = React.useState<IFluidClientDebugger | undefined>(
```
Nit: Might be worth renaming `clientDebugger` to `selectedClientDebugger` or something like that.

chentong7 replied:
Renamed

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx:
```diff
@@ -32,48 +39,68 @@ export interface FluidClientDebuggerProps {
+            debuggerRegistry.off("debuggerClosed", onDebuggerChanged);
         };
-    }, [clientDebugger, setIsContainerDisposed]);
+    }, [clientDebugger, clientDebuggers, setIsContainerDisposed, setClientDebuggers]);
```
We need to add `debuggerRegistry` to the list of dependencies here. Otherwise if it changes, this effect won't re-run.

It also doesn't look like we're using `clientDebuggers` anywhere in this effect, so it can probably be removed from the deps list.

chentong7 replied:
I see! Fixed it. Thx!

Josmithr replied:
Looks like it's still in the dependencies list :)

chentong7 commented on packages/tools/client-debugger/client-debugger-view/src/CommonProps.ts:
```diff
@@ -9,6 +9,18 @@ import { IFluidClientDebugger } from "@fluid-tools/client-debugger";
+/**
+ * Base interface for component props that observe the Client debuggers list update.
+ *
+ * @internal
```
I see. Then I think I should remove this tag now. If there are external consumers, I will add it back to index.ts at that point. Thx

chentong7 commented on packages/tools/client-debugger/client-debugger/src/Registry.ts:
```diff
@@ -14,35 +16,103 @@ import { IFluidClientDebugger } from "./IFluidClientDebugger";
+		super();
+	}
+
+	public initializeDebugger(props: FluidClientDebuggerProps): void {
```
Added doc

chentong7 committed:
update PR

chentong7 committed:
Merge branch 'main' of https://github.com/chentong7/FluidFramework into tong_pr

chentong7 committed:
Delete Registry.ts

chentong7 committed:
Revert "Delete Registry.ts"

This reverts commit d98b3e6e1e6b9ad25e0fc224373ab1a4308c913c.

chentong7 committed:
Delete Registry.ts

chentong7 committed:
testing in local

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerSelectionDropdown.tsx:
```diff
@@ -0,0 +1,80 @@
+/**
+ * {@link ContainerSelectionDropdownProps} input props.
+ *
+ * @internal
```
Nit: so long as we don't export this in the root `index.ts` file of this package, we don't need the `@internal` flag here.

chentong7 replied:
Removed

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerSelectionDropdown.tsx:
```diff
@@ -0,0 +1,80 @@
+}
+
+/**
+ * A dropdown that displays all registered containers.
```
I would recommend documenting this in terms of its use in selecting a container from a list. E.g. something like
```suggestion
 * A dropdown menu for selecting the Fluid Container to display debug information about.
```

chentong7 replied:
Thank you for suggestion. Replace the comments

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerSelectionDropdown.tsx:
```diff
@@ -0,0 +1,80 @@
+        option?: IDropdownOption,
+    ): void => {
+        if (option) {
+            const selectedDebugger = clientDebuggers.find((c) => {
```
Nit: for accessibility reasons, I always recommend against single-letter variables like `c`. I would recommend `debugger` or `clientDebugger` or something more descriptive instead :)

chentong7 replied:
Renamed it!

Josmithr replied:
Thanks!

Josmithr commented on packages/tools/client-debugger/client-debugger/src/test/ClientDebugger.test.ts:
```diff
@@ -11,7 +11,7 @@ import { IFluidClientDebugger } from "../IFluidClientDebugger";
 import {
```
I would recommend adding a test that uses the registry's listener pattern, and verifies that the listeners get called when we register / close debugger instances.

chentong7 replied:
Will appear after the unit tests PR merge into this one.

Josmithr reviewed (commented):
This is looking very close! I left a few style comments, and I do think we should add a simple unit test for the DebuggerRegistry's listener hooks, but otherwise I think this looks great!

chentong7 committed:
test locally together

chentong7 committed:
Adjust regarding to the PR comments

chentong7 committed:
test again

chentong7 committed:
Merge branch 'main' of https://github.com/chentong7/FluidFramework into tong_pr

chentong7 committed:
test again

chentong7 committed:
add unit tests

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerSelectionDropdown.tsx:
```diff
@@ -0,0 +1,79 @@
+        if (option) {
+            const selectedDebugger = clientDebuggers.find((clientDebugger) => {
+                return clientDebugger.containerId === (option.key as string);
+            }) as IFluidClientDebugger;
```
Rather than doing a type-cast here, I think it would be better to check if the value returned from `find` is `undefined`. If it is, we have an internal logic bug, and we should probably throw an error.

Something like
```suggestion
            });
            if (selectedDebugger === undefined) {
                throw new Error(`Could not find a debugger associated with Container ID "${option.key}". This likely indicates an internal state issue.`);
            }
```

Josmithr reviewed (changes requested):
There are just a couple of remaining issues I see, and a few nit-picky suggestions. Otherwise I think this is looking great!

chentong7 commented on packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx:
```diff
@@ -25,79 +29,102 @@ export interface FluidClientDebuggerProps {
 
+// Drop down + single container view
 /**
  * Renders the debug view for an active debugger session registered using
```
Re-word the comment

chentong7 commented on packages/tools/client-debugger/client-debugger/src/test/ClientDebugger.test.ts:
```diff
@@ -23,45 +23,68 @@ describe("ClientDebugger unit tests", () => {
 
     it("Closing debugger removes it from global (window) registry and disposes it.", () => {
-        const clientDebugger = initializeDebugger();
+        let clientDebugger = initializeDebugger(containerId, container!);
```
Modified to have two separate variables.

chentong7 commented on packages/tools/client-debugger/client-debugger/src/Registry.ts:
```diff
@@ -111,7 +172,7 @@ export function getFluidClientDebuggers(): IFluidClientDebugger[] {
 
     const clientDebuggers: IFluidClientDebugger[] = [];
-    for (const [, clientDebugger] of debuggerRegistry) {
+    for (const [, clientDebugger] of debuggerRegistry.getRegisteredDebuggers()) {
```
Hmm... I think I would prefer to just have a public function to call to  get the list of debuggers without need to call getRegistery() first.

chentong7 commented on packages/tools/client-debugger/client-debugger-view/src/test/app/App.tsx:
```diff
@@ -183,17 +206,52 @@ const appViewPaneStackStyles = mergeStyles({
+        </Stack>
+    );
+
+    // TODO: separate loading spinner for each container as it is async loaded
```
Removed

chentong7 commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerSelectionDropdown.tsx:
```diff
@@ -0,0 +1,79 @@
+        event: React.FormEvent<HTMLDivElement>,
+        option?: IDropdownOption,
+    ): void => {
+        if (option) {
```
Fixed. I though it's a common expression for ts

chentong7 commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerSelectionDropdown.tsx:
```diff
@@ -0,0 +1,79 @@
+        const options: IDropdownOption[] = [];
+        for (const each_debugger of debuggers) {
+            if (each_debugger.containerId === containerId) {
+                selectedKey = each_debugger.containerNickname ?? each_debugger.containerId;
```
You are right. We never have a nickName as key.  We don't need this if-clause at all.

chentong7 commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerSelectionDropdown.tsx:
```diff
@@ -0,0 +1,79 @@
+
+    const { clientDebuggers, containerId } = props;
+
+    let selectedKey = containerId;
```
Removed it! We don't need the `selectedKey` at all.

chentong7 committed:
update pr

chentong7 committed:
update Pr

chentong7 committed:
Update packages/tools/client-debugger/client-debugger-view/src/components/ContainerSelectionDropdown.tsx

Co-authored-by: Joshua Smithrud <54606601+Josmithr@users.noreply.github.com>

chentong7 committed:
Update packages/tools/client-debugger/client-debugger-view/src/components/ContainerSelectionDropdown.tsx

Co-authored-by: Joshua Smithrud <54606601+Josmithr@users.noreply.github.com>

chentong7 commented on packages/tools/client-debugger/client-debugger-view/src/components/data-object-views/SharedStringView.tsx:
```diff
@@ -23,10 +23,16 @@ export interface SharedStringViewProps {
 export function SharedStringView(props: SharedStringViewProps): React.ReactElement {
     const { sharedString } = props;
 
+    const [text, setText] = React.useState<SharedString>(sharedString);
```
switch container will not update the shareString here. This fille will be changed after this task: https://dev.azure.com/fluidframework/internal/_workitems/edit/2659/

chentong7 commented on packages/tools/client-debugger/client-debugger-view/src/index.ts:
```diff
@@ -29,7 +29,10 @@ export {
-export { FluidClientDebugger, FluidClientDebuggerProps } from "./Debugger";
+export {
+    FluidClientDebuggers as FluidClientDebugger,
+    FluidClientDebuggersProps as FluidClientDebuggerProps,
```
Fixed

chentong7 commented on packages/tools/client-debugger/client-debugger-view/src/test/app/App.tsx:
```diff
@@ -90,8 +90,13 @@ async function populateRootMap(container: IFluidContainer): Promise<void> {
 /**
  * React hook for asynchronously creating / loading the Fluid Container.
```
Added

chentong7 commented on packages/tools/client-debugger/client-debugger-view/src/components/ContainerSelectionDropdown.tsx:
```diff
@@ -0,0 +1,79 @@
+    props: ContainerSelectionDropdownProps,
+): React.ReactElement {
+    const dropdownStyles: Partial<IDropdownStyles> = {
+        dropdown: { width: "300px", zIndex: "1" },
```
It will generate a fixed size pop-up dropdown when you click on it but resize along with debug panel when not using it.

chentong7 replied:
I'm trying to make a screenshot to show it. Still fixing my `npm`

chentong7 replied:
![1](https://user-images.githubusercontent.com/114451900/215631598-c09054be-4dfc-4525-b233-a563ccf96f3c.png)
![2](https://user-images.githubusercontent.com/114451900/215631655-612f6caa-72b7-4738-9f81-992405952c24.png)

Josmithr replied:
Looks good to me!

chentong7 committed:
testing local

chentong7 committed:
Merge branch 'main' of https://github.com/chentong7/FluidFramework into tong_pr

chentong7 committed:
test working

chentong7 committed:
Update validateTestClientUtilsPrevious.generated.ts

chentong7 committed:
Update validateTestClientUtilsPrevious.generated.ts

chentong7 committed:
Fix format

chentong7 committed:
Merge branch 'tong_pr' of https://github.com/chentong7/FluidFramework into tong_pr

chentong7 committed:
Merge branch 'main' of https://github.com/chentong7/FluidFramework into tong_pr

chentong7 committed:
better way to fix webpack

chentong7 commented on packages/tools/client-debugger/client-debugger-view/webpack.config.js:
```diff
@@ -47,7 +47,7 @@ module.exports = (env) => {
 			plugins: [
 				new webpack.ProvidePlugin({
-					process: "process/browser",
+					process: require.resolve("process/browser"),
```
This is to fix test class can't find /node_modules. By configuring this in webpack, `process` will be required to resolve.

Sample log:
Field 'browser' doesn't contain a valid alias configuration
    resolve as module
      /home/tong/workspace/FluidFramework/packages/tools/client-debugger/client-debugger-view/src/test/node_modules doesn't exist or is not a directory
      /home/tong/workspace/FluidFramework/packages/tools/client-debugger/client-debugger-view/src/node_modules doesn't exist or is not a directory
      looking for modules in /home/tong/workspace/FluidFramework/packages/tools/client-debugger/client-debugger-view/node_modules
        /home/tong/workspace/FluidFramework/packages/tools/client-debugger/client-debugger-view/node_modules/process doesn't exist
      /home/tong/workspace/FluidFramework/packages/tools/client-debugger/node_modules doesn't exist or is not a directory
      /home/tong/workspace/FluidFramework/packages/tools/node_modules doesn't exist or is not a directory
      /home/tong/workspace/FluidFramework/packages/node_modules doesn't exist or is not a directory
      looking for modules in /home/tong/workspace/FluidFramework/node_modules
        /home/tong/workspace/FluidFramework/node_modules/process doesn't exist
      /home/tong/workspace/node_modules doesn't exist or is not a directory
      /home/tong/node_modules doesn't exist or is not a directory
      /home/node_modules doesn't exist or is not a directory
      /node_modules doesn't exist or is not a directory
@ ./src/test/app/App.tsx 27:26-55

chentong7 committed:
Update webpack.config.js

chentong7 commented on packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx:
```diff
@@ -25,79 +29,102 @@ export interface FluidClientDebuggerProps {
 	renderOptions?: RenderOptions;
 }
 
+// Drop down to show 2+ containers and manage the selected container in debugger view.
```
Moved it.

chentong7 committed:
test locally

chentong7 committed:
Merge branch 'tong_pr' of https://github.com/chentong7/FluidFramework into tong_pr

chentong7 committed:
better resize

Josmithr reviewed (approved):
🚀 Looks good!!
